In [8]:
import random
import numpy as np

def swap_rows(matrix, row1, row2):
    """Swaps two rows of a matrix."""
    matrix[row1], matrix[row2] = matrix[row2], matrix[row1]

def scale_row(matrix, row, scalar):
    """Scales a row of a matrix by a scalar value."""
    for i in range(len(matrix[row])):
        matrix[row][i] *= scalar

def add_multiple_of_row_to_row(matrix, source_row, target_row, scalar):
    """Adds a multiple of one row to another row in a matrix."""
    for i in range(len(matrix[target_row])):
        matrix[target_row][i] += scalar * matrix[source_row][i]

def ref(matrix):
    """Converts a matrix to row echelon form (REF) with pivot normalization."""
    num_rows = len(matrix)
    num_cols = len(matrix[0])
    lead = 0

    for r in range(num_rows):
        if lead >= num_cols:
            break

        i = r
        while matrix[i][lead] == 0:
            i += 1
            if i == num_rows:
                i = r
                lead += 1
                if lead == num_cols:
                    break

        swap_rows(matrix, i, r)  # Pivoting

        # Normalize the pivot element to 1
        pivot_value = matrix[r][lead]
        # Divide each element in the pivot row by the pivot value
        for i in range(len(matrix[r])):  # Iterate through each element in the row
            matrix[r][i] /= pivot_value

        for j in range(r + 1, num_rows):
            if matrix[j][lead] != 0:
                add_multiple_of_row_to_row(matrix, r, j, -matrix[j][lead])  # Elimination (no division needed now)

        lead += 1

def rref(matrix):
    """Converts a matrix to reduced row echelon form (RREF)."""
    ref(matrix)  # Perform REF first
    num_rows = len(matrix)
    num_cols = len(matrix[0])

    for r in range(num_rows - 1, -1, -1):  # Iterate from bottom to top
        for j in range(r):
            if matrix[j][r] != 0:
                add_multiple_of_row_to_row(matrix, r, j, -matrix[j][r])  # Eliminate elements above the pivot

        if matrix[r][r] != 0:
            scale_row(matrix, r, 1 / matrix[r][r])  # Scale the pivot to 1

def identify_pivot_columns(matrix):
    """Identifies pivot and non-pivot columns in a matrix."""
    pivot_cols = []
    for i in range(len(matrix)):
        for j in range(len(matrix[0])):
            if matrix[i][j] == 1 and all(matrix[k][j] == 0 for k in range(i)):  # Pivot condition
                pivot_cols.append(j)
                break
    non_pivot_cols = [i for i in range(len(matrix[0])) if i not in pivot_cols]
    return pivot_cols, non_pivot_cols

def find_particular_solution(rref_matrix):
    """Finds the particular solution to Ax = b from the RREF matrix."""
    solution = [0] * (len(rref_matrix[0]) - 1)  # Initialize solution vector
    print("solution ",solution)
    for i in range(len(rref_matrix)):
        if 1 in rref_matrix[i]:  # Check if there's a leading 1
            solution[rref_matrix[i].index(1)] = rref_matrix[i][-1]
        else:
            # Handle cases without a leading 1 (e.g., dependent variables)
            print("Row without a leading 1:", rref_matrix[i])
            # You might want to assign a symbolic value or raise a specific exception here
    return solution

def get_homogeneous_solutions(matrix):
    rows = len(matrix)
    cols = len(matrix[0])
    pivot_columns, nonpivot_columns = identify_pivot_columns(matrix)

    solutions = []

    for col in nonpivot_columns:
        solution = np.zeros((cols - 1, 1))
        if col < len(solution):  # Check if col is within bounds
            solution[col] = 1

        for row, pivot_col in enumerate(pivot_columns):
            if pivot_col < len(solution):  # Check if pivot_col is within bounds
                solution[pivot_col, 0] = -matrix[row][col]

        solutions.append(solution.flatten().tolist())

    return solutions


    
while True:
    try:
        m = int(input("Enter the number of rows (m) (must be less than the number of columns): "))
        n = int(input("Enter the number of columns (n): "))
        if m > n:
            print("Invalid input. Rows (m) must be less than columns (n). Please try again.")
            continue
        break
    except ValueError:
        print("Invalid input. Please enter integers for rows and columns.")

while True:
    choice = input("Choose an option for matrix A:\n1. Generate a random matrix\n2. Enter matrix elements manually\nEnter your choice (1 or 2): ")
    if choice in ["1", "2"]:
        break
    else:
        print("Invalid choice. Please enter 1 or 2.")

if choice == "1":
    A = [[random.randint(1, 10) for _ in range(n)] for _ in range(m)]  # Generate random matrix A
else:
    A = []
    for i in range(m):
        row = []
        for j in range(n):
            while True:
                try:
                    value = int(input(f"Enter value for element A({i+1},{j+1}): "))
                    row.append(value)
                    break
                except ValueError:
                    print("Invalid input. Please enter an integer.")
        A.append(row)

# Input for vector b
b = []
for i in range(m):
    while True:
        try:
            value = int(input(f"Enter value for element b{i+1}: "))
            b.append(value)
            break
        except ValueError:
            print("Invalid input. Please enter an integer.")

# Construct augmented matrix
augmented_matrix = [[A[i][j] for j in range(n)] + [b[i]] for i in range(m)]

print("\nMatrix A \n")
for row in A:
    print([int(x) for x in row])

print("\nVector b \n",b)


print("\nAugmented matrix:")
for row in augmented_matrix:
    print([int(x) for x in row])

# Perform REF
ref(augmented_matrix)
print("\nREF:")
for row in augmented_matrix:
    print([int(x) for x in row])

# Perform RREF
rref(augmented_matrix)
print("\nRREF:")
for row in augmented_matrix:
    print([int(x) for x in row])
    

# Identify pivot and non-pivot columns
pivot_cols, non_pivot_cols = identify_pivot_columns(augmented_matrix)
print("\nPivot columns:", pivot_cols)
print("Non-pivot columns:", non_pivot_cols)

# Find particular solution
particular_solution = find_particular_solution(augmented_matrix)
print("\nParticular solution:", particular_solution)

# Find solutions to Ax = 0
# Get solutions to Ax = 0
homogeneous_solutions = get_homogeneous_solutions(augmented_matrix)
print("\nSolutions to Ax = 0:")
for solution in homogeneous_solutions:
    print(solution)

# Construct and print the general solution
general_solution = [particular_solution] + homogeneous_solutions
print("\nGeneral Solution:")
for solution in general_solution:
    print(solution)


Matrix A 

[2, 6, 8, 3, 5, 4, 7]
[1, 3, 8, 1, 6, 2, 10]
[3, 8, 7, 4, 5, 5, 6]
[2, 7, 3, 6, 1, 7, 9]
[3, 7, 5, 1, 5, 1, 9]

Vector b 
 [-1, 2, 3, -2, 4]

Augmented matrix:
[2, 6, 8, 3, 5, 4, 7, -1]
[1, 3, 8, 1, 6, 2, 10, 2]
[3, 8, 7, 4, 5, 5, 6, 3]
[2, 7, 3, 6, 1, 7, 9, -2]
[3, 7, 5, 1, 5, 1, 9, 4]

REF:
[1, 3, 4, 1, 2, 2, 3, 0]
[0, 1, 5, 0, 2, 1, 4, -4]
[0, 0, 1, 0, 0, 0, 1, 0]
[0, 0, 0, 1, 1, 1, 11, 7]
[0, 0, 0, 0, 1, 0, 7, 3]

RREF:
[1, 0, 0, 0, 0, 0, -26, 9]
[0, 1, 0, 0, 0, 0, 11, -4]
[0, 0, 1, 0, 0, 0, -4, -1]
[0, 0, 0, 1, 0, 1, -1, 2]
[0, 0, 0, 0, 1, 0, 7, 3]

Pivot columns: [0, 1, 2, 3, 4]
Non-pivot columns: [5, 6, 7]
solution  [0, 0, 0, 0, 0, 0, 0]

Particular solution: [9.108108108108107, -4.594594594594595, -1.7297297297297294, 2.3513513513513518, 3.0270270270270268, 0, 0]

Solutions to Ax = 0:
[-0.4324324324324318, 0.37837837837837807, -0.08108108108108099, -1.4054054054054053, -0.1081081081081082, 1.0, 0.0]
[26.891891891891888, -11.405405405405403, 4.729729729729729, 1.6486

<h1> Amit Bhai </h1>
Matrix A 

[6, 3, 4, 8, 10, 4, 10]
[6, 3, 8, 9, 1, 6, 3]
[2, 7, 3, 5, 6, 4, 5]
[4, 4, 7, 6, 10, 3, 10]
[2, 10, 8, 5, 6, 4, 9]

Vector b 
 [2, 1, 4, 5, -2]

Augmented matrix:
[6, 3, 4, 8, 10, 4, 10, 2]
[6, 3, 8, 9, 1, 6, 3, 1]
[2, 7, 3, 5, 6, 4, 5, 4]
[4, 4, 7, 6, 10, 3, 10, 5]
[2, 10, 8, 5, 6, 4, 9, -2]

REF:
[1, 0, 0, 1, 1, 0, 1, 0]
[0, 1, 0, 0, 0, 0, 0, 0]
[0, 0, 1, 0, -2, 0, -1, 0]
[0, 0, 0, 1, -10, 2, -8, -3]
[0, 0, 0, 0, 1, 0, 0, 0]

RREF:
[1, 0, 0, 0, 0, 0, 3, -8]
[0, 1, 0, 0, 0, 0, 0, -2]
[0, 0, 1, 0, 0, 0, 0, 0]
[0, 0, 0, 1, 0, 1, -2, 6]
[0, 0, 0, 0, 1, 0, 0, 0]

Pivot columns: [0, 1, 2, 3, 4]
Non-pivot columns: [5, 6, 7]
solution  [0, 0, 0, 0, 0, 0, 0]

Particular solution: [-8.74712134632418, -2.429583702391496, 0.25775022143489745, 6.418954827280779, 0.9388839681133744, 0, 0]

Solutions to Ax = 0:
[0.6576616474756422, -0.0664304694419841, 0.039858281665190454, -1.1310894596988486, 0.11426040744021258, 1.0, 0.0]
[-3.325952170062002, -0.8963684676705048, -0.2621789193976971, 2.4844995571302038, -0.6182462356067316, 0.0, 1.0]
[8.74712134632418, 2.429583702391496, -0.25775022143489745, -6.418954827280779, -0.9388839681133744, 0.0, 0.0]

General Solution:
[-8.74712134632418, -2.429583702391496, 0.25775022143489745, 6.418954827280779, 0.9388839681133744, 0, 0]
[0.6576616474756422, -0.0664304694419841, 0.039858281665190454, -1.1310894596988486, 0.11426040744021258, 1.0, 0.0]
[-3.325952170062002, -0.8963684676705048, -0.2621789193976971, 2.4844995571302038, -0.6182462356067316, 0.0, 1.0]
[8.74712134632418, 2.429583702391496, -0.25775022143489745, -6.418954827280779, -0.9388839681133744, 0.0, 0.0]

<h1> Vishal </h1>
Matrix A 

[9, 10, 8, 4, 2, 1, 4]
[6, 8, 9, 6, 2, 3, 4]
[10, 2, 8, 3, 5, 5, 10]
[3, 7, 2, 8, 6, 4, 1]
[7, 8, 4, 6, 9, 3, 6]

Vector b 
 [1, 4, -5, 6, 3]

Augmented matrix:
[9, 10, 8, 4, 2, 1, 4, 1]
[6, 8, 9, 6, 2, 3, 4, 4]
[10, 2, 8, 3, 5, 5, 10, -5]
[3, 7, 2, 8, 6, 4, 1, 6]
[7, 8, 4, 6, 9, 3, 6, 3]

REF:
[1, 1, 0, 0, 0, 0, 0, 0]
[0, 1, 2, 2, 0, 1, 1, 2]
[0, 0, 1, 0, 0, 0, 0, 0]
[0, 0, 0, 1, 0, 0, 0, 0]
[0, 0, 0, 0, 1, 0, 0, 0]

RREF:
[1, 0, 0, 0, 0, 0, 0, -1]
[0, 1, 0, 0, 0, 0, 0, 0]
[0, 0, 1, 0, 0, 0, 0, 0]
[0, 0, 0, 1, 0, 0, 0, 0]
[0, 0, 0, 0, 1, 0, 0, 0]

Pivot columns: [0, 1, 2, 3, 4]
Non-pivot columns: [5, 6, 7]
solution  [0, 0, 0, 0, 0, 0, 0]

Particular solution: [-1.1268415883260836, 0.6680230110846073, 0.3384313175249056, 0.3085449698330292, 0.2598568822786587, 0, 0]

Solutions to Ax = 0:
[-0.31892802020485433, 0.5745755577381781, -0.04728497263925924, -0.8801739862494734, 0.011786165286936844, 1.0, 0.0]
[-0.251157569804967, 0.3913287498246105, -0.7015574575557737, 0.36537112389504695, -0.7509471025677001, 0.0, 1.0]
[1.1268415883260836, -0.6680230110846073, -0.3384313175249056, -0.3085449698330292, -0.2598568822786587, 0.0, 0.0]

General Solution:
[-1.1268415883260836, 0.6680230110846073, 0.3384313175249056, 0.3085449698330292, 0.2598568822786587, 0, 0]
[-0.31892802020485433, 0.5745755577381781, -0.04728497263925924, -0.8801739862494734, 0.011786165286936844, 1.0, 0.0]
[-0.251157569804967, 0.3913287498246105, -0.7015574575557737, 0.36537112389504695, -0.7509471025677001, 0.0, 1.0]
[1.1268415883260836, -0.6680230110846073, -0.3384313175249056, -0.3085449698330292, -0.2598568822786587, 0.0, 0.0]

<h1> Anshuman </h1>

Matrix A 

[2, 6, 8, 3, 5, 4, 7]
[1, 3, 8, 1, 6, 2, 10]
[3, 8, 7, 4, 5, 5, 6]
[2, 7, 3, 6, 1, 7, 9]
[3, 7, 5, 1, 5, 1, 9]

Vector b 
 [-1, 2, 3, -2, 4]

Augmented matrix:
[2, 6, 8, 3, 5, 4, 7, -1]
[1, 3, 8, 1, 6, 2, 10, 2]
[3, 8, 7, 4, 5, 5, 6, 3]
[2, 7, 3, 6, 1, 7, 9, -2]
[3, 7, 5, 1, 5, 1, 9, 4]

REF:
[1, 3, 4, 1, 2, 2, 3, 0]
[0, 1, 5, 0, 2, 1, 4, -4]
[0, 0, 1, 0, 0, 0, 1, 0]
[0, 0, 0, 1, 1, 1, 11, 7]
[0, 0, 0, 0, 1, 0, 7, 3]

RREF:
[1, 0, 0, 0, 0, 0, -26, 9]
[0, 1, 0, 0, 0, 0, 11, -4]
[0, 0, 1, 0, 0, 0, -4, -1]
[0, 0, 0, 1, 0, 1, -1, 2]
[0, 0, 0, 0, 1, 0, 7, 3]

Pivot columns: [0, 1, 2, 3, 4]
Non-pivot columns: [5, 6, 7]
solution  [0, 0, 0, 0, 0, 0, 0]

Particular solution: [9.108108108108107, -4.594594594594595, -1.7297297297297294, 2.3513513513513518, 3.0270270270270268, 0, 0]

Solutions to Ax = 0:
[-0.4324324324324318, 0.37837837837837807, -0.08108108108108099, -1.4054054054054053, -0.1081081081081082, 1.0, 0.0]
[26.891891891891888, -11.405405405405403, 4.729729729729729, 1.6486486486486474, -7.027027027027026, 0.0, 1.0]
[-9.108108108108107, 4.594594594594595, 1.7297297297297294, -2.3513513513513518, -3.0270270270270268, 0.0, 0.0]

General Solution:
[9.108108108108107, -4.594594594594595, -1.7297297297297294, 2.3513513513513518, 3.0270270270270268, 0, 0]
[-0.4324324324324318, 0.37837837837837807, -0.08108108108108099, -1.4054054054054053, -0.1081081081081082, 1.0, 0.0]
[26.891891891891888, -11.405405405405403, 4.729729729729729, 1.6486486486486474, -7.027027027027026, 0.0, 1.0]
[-9.108108108108107, 4.594594594594595, 1.7297297297297294, -2.3513513513513518, -3.0270270270270268, 0.0, 0.0]